substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/PreMainSupport.java

```diff
@@ -0,0 +1,231 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.core;
+
+import com.oracle.svm.core.util.VMError;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+
+import java.lang.instrument.ClassDefinition;
+import java.lang.instrument.ClassFileTransformer;
+import java.lang.instrument.Instrumentation;
+import java.lang.instrument.UnmodifiableClassException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.jar.JarFile;
+
+/**
+ * Java agent can do instrumentation initialization work in premain phase. This class supports
+ * registering such premain methods at native image build time and invoking them at native image
+ * runtime. <br>
+ * JVM supports two kind of premain methods:
+ * <ol>
+ * <li>{@code public static void premain(String agentArgs, Instrumentation inst)}</li>
+ * <li>{@code public static void premain(String agentArgs)}</li>
+ * </ol>
+ * For the first one, at registration time we will set the second parameter with an instance of
+ * {@link NativeImageNoOpRuntimeInstrumentation} class which does not do any actual work as no
+ * instrumentation can do in native code at runtime.
+ * <p>
+ * <b>Be noticed</b>, the original agent premain method may not work well at native image runtime
+ * even if the input {@link Instrumentation} class is replaced with
+ * {@link NativeImageNoOpRuntimeInstrumentation}.
+ * </p>
+ * <p>
+ * It is the agent developers' responsibility to implement a native version of their agent premain
+ * method. It can be implemented in two ways:
+ * <ul>
+ * <li>Isolate code by checking current runtime. For example: <code>
+ *         <pre>
+ *         if (ImageInfo.inImageRuntimeCode()) {
```

Comment:
I would prefer to use the system property directly here, or at least mention it. This will not require extra dependencies to use this feature.

Reply:
Fixed